api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java

```diff
@@ -358,6 +360,71 @@ private void addConceptsToSet(
     return ResponseEntity.ok(toClientConceptSet(dbConceptSet));
   }
 
+  @Override
+  public ResponseEntity<ConceptSet> copyConceptSet(
+      String fromWorkspaceNamespace,
+      String fromWorkspaceId,
+      String fromConceptSetId,
+      CopyRequest copyRequest) {
+    Timestamp now = new Timestamp(clock.instant().toEpochMilli());
+    workspaceService.enforceWorkspaceAccessLevel(
+        fromWorkspaceNamespace, fromWorkspaceId, WorkspaceAccessLevel.READER);
+    Workspace toWorkspace =
+        workspaceService.get(
+            copyRequest.getToWorkspaceNamespace(), copyRequest.getToWorkspaceName());
+    Workspace fromWorkspace = workspaceService.get(fromWorkspaceNamespace, fromWorkspaceId);
+    workspaceService.getWorkspaceEnforceAccessLevelAndSetCdrVersion(
+        toWorkspace.getWorkspaceNamespace(),
+        toWorkspace.getFirecloudName(),
+        WorkspaceAccessLevel.WRITER);
+    if (toWorkspace.getCdrVersion().getCdrVersionId()
+        != fromWorkspace.getCdrVersion().getCdrVersionId()) {
+      throw new BadRequestException(
+          "New workspace does not have the same CDR version as current workspace");
+    }
+    org.pmiops.workbench.db.model.ConceptSet conceptSet =
+        conceptSetDao.findOne(Long.valueOf(fromConceptSetId));
+    if (conceptSet == null) {
+      throw new NotFoundException(
+          String.format(
+              "Concept set %s does not exist",
+              generateConceptSetPathFromWorkspaceAndIdentifier(
+                  fromWorkspaceNamespace, fromWorkspaceId, fromConceptSetId)));
+    }
+    org.pmiops.workbench.db.model.ConceptSet newConceptSet =
+        new org.pmiops.workbench.db.model.ConceptSet(conceptSet);
+
+    newConceptSet.setName(copyRequest.getNewName());
+    newConceptSet.setCreator(userProvider.get());
+    newConceptSet.setWorkspaceId(toWorkspace.getWorkspaceId());
+    newConceptSet.setCreationTime(now);
+    newConceptSet.setLastModifiedTime(now);
+    newConceptSet.setVersion(INITIAL_VERSION);
+
+    try {
+      newConceptSet = conceptSetDao.save(newConceptSet);
+      userRecentResourceService.updateConceptSetEntry(
+          toWorkspace.getWorkspaceId(),
+          userProvider.get().getUserId(),
+          newConceptSet.getConceptSetId(),
+          now);
+    } catch (DataIntegrityViolationException e) {
+      throw new ConflictException(
+          String.format(
+              "Concept set %s already exists.",
```

Comment:
I would much rather us try to figure out what other error cases might exist, and handle all of them. This is currently (as far as I know) the only cause that we can see from this, because it is copying a valid concept set and changing the name only. Can you think of any other errors we might see?